ARROW-13623 [R] write_csv_arrow update to follow the signature of readr::write_csv

dragosmg opened the pull request:
@thisisnic @jonkeane & @paleolimbot This is my first attempt at the task. Not convinced this is the best approach for exposing the args to the user (in prep for hooking them up to the C++ functionality). 
I plan on adding more unit test for this PR.
I'd be tempted to suggest we start using some sort of `lifecycle` approach. In such a case, we could have a function called `in_progress()` which would indicate we are working on exposing functionality. Similarly the `tidyverse` uses, for example, `lifecycle::deprecated()`

thisisnic commented:
Did you say you'd opened some JIRA tickets for implementing the missing functionality? If so, how about popping links to those in some comments, and then it'll be more apparent what the dependencies are for implementing the new functionality?

dragosmg commented:
No tickets just yet. It's not 100% clear how close we want to end up to the `readr::write_csv()` signature - e.g. probably we won't be implementing `num_threads`.

jonkeane commented on r/R/csv.R:
```diff
@@ -623,8 +632,38 @@ readr_to_csv_convert_options <- function(na,
                             include_header = TRUE,
                             batch_size = 1024L) {
+
+  passed_args <- as.list(match.call()[-1])
+
+  arrow_write_opts <- names(formals(CsvWriteOptions$create))
+  write_args <- names(formals(write_csv_arrow))
+
+  unsupported_opts <- setdiff(
+    write_args,
+    union(arrow_write_opts, c("x", "sink"))
+  )
+
+  unsupported_passed_args <- names(unlist(passed_args[unsupported_opts]))
+
+  if (length(unsupported_passed_args)) {
+    stop(
+      "The following ",
+      ngettext(length(unsupported_passed_args), "argument is ", "arguments are "),
+      "not yet supported in Arrow: ",
+      oxford_paste(unsupported_passed_args),
+      call. = FALSE
+    )
+  }
```
Yeah, I like this approach proposed by @paleolimbot here. It should also avoid any need for parsing the formals and comparing those.

You can also make a helper function for that [like we do with the reading side](readr_to_csv_parse_options)

dragosmg replied:
The link is broken, but I think I know what you mean. That was one of my other possible directions. Cool. Off to do some work and will update the PR.

dragosmg replied:
@jonkeane & @paleolimbot do we change the name of the write_csv_arrow() argument from sink to file? What about include_header to col_names?

jonkeane replied:
Yeah `sink` is a lot more than file path, and matches our other functions. How awkward would it be to support both sink and filename? It's a bit awkward, but it would make this match readr while also maintaining our arrow patterns.

dragosmg replied:
Done. @jonkeane please have a look to see if what I did makes sense. I think we could improve the arg description and maybe not `message()` about the whole `sink` `file` situation?

jonkeane commented on r/tests/testthat/test-csv.R:
```diff
@@ -378,3 +378,10 @@ test_that("time mapping work as expected (ARROW-13624)", {
+    write_csv_arrow(tbl, csv_file, append = FALSE),
+    "The following argument is not yet supported in Arrow: \"append\""
+  )
+})
```
Thanks for proactively adding the error test, that's great!

jonkeane reviewed (changes requested):
A few comments, generally I agree with @paleolimbot about the arguments / flagging those that aren't (yet) supported

dragosmg commented on r/R/csv.R:
```diff
@@ -612,6 +622,8 @@ readr_to_csv_convert_options <- function(na,
 #' system (`SubTreeFileSystem`)
 #' @param include_header Whether to write an initial header line with column names
 #' @param batch_size Maximum number of rows processed at a time. Default is 1024.
+#' @param write_options read_options see [file reader options][CsvWriteOptions]
```
The downside to this is that we do not have much in terms of content for `CsvWriteOptions`. We direct to it from a couple of places, but maybe we need to extent the documentation, more so once we surface more options from the C++ code.

dragosmg commented:
@paleolimbot & @jonkeane I think the current state of the PR addresses your suggestions. The only thing we need to decide on is whether to keep the current names for `include_header` and `sink`. I think we should. What do you think?

jonkeane commented on r/R/csv.R:
```diff
@@ -624,8 +636,27 @@ readr_to_csv_convert_options <- function(na,
+                            batch_size = 1024L,
+                            write_options = NULL,
+                            ...) {
+
```
```suggestion
```

dragosmg replied:
Done

jonkeane reviewed (changes requested):
This is looking good, one small suggestion.

> The only thing we need to decide on is whether to keep the current names for include_header and sink. I think we should. What do you think?

If we can do both `sink` (to match our other functions, and keep the extra kinds of objects we support there) and `filename` that would be fantastic (though we should be clear in our docs that they are controlling ~the same thing so someone doesn't try to or think that they need to supply `write_csv_arrow(df, sink = "foo.csv", filename = "bar.csv")` Same thing with `include_header` and `col_names` if we make it clear that they are ~the same, we are supplying `col_names` so this can be (hopefully, though see below) a copy/paste replacement that would be good.

Is there already a ira (or multiple liras) for the arguments that we don't support that we want to?

dragosmg commented:
I have created a series of follow-up Jiras:
* [ARROW-14902](https://issues.apache.org/jira/browse/ARROW-14902) - umbrella ticket to update `write_csv_arrow()` once the C++ functionality is there.
* [ARROW-14903](https://issues.apache.org/jira/browse/ARROW-14903) - control the string used for missing data
* [ARROW-14904](https://issues.apache.org/jira/browse/ARROW-14904) - enable append/overwrite
* [ARROW-14905](https://issues.apache.org/jira/browse/ARROW-14905) - control quoting
* [ARROW-14906](https://issues.apache.org/jira/browse/ARROW-14906) - control the type of escape used for quoting
* [ARROW-14907](https://issues.apache.org/jira/browse/ARROW-14907) - control the end-of-line character

ianmcook commented on r/R/csv.R:
```diff
@@ -623,9 +638,58 @@ readr_to_csv_convert_options <- function(na,
+  }
+
+  if (!missing(file) && !missing(sink)) {
+    stop(
+      paste("You have supplied both \"file\" and \"sink\" arguments. Please",
+            "supply only one of them."),
+      call. = FALSE
+    )
```
Can simplify this by removing the `paste()` since `stop()` already pastes together its arguments (with no separator)
```suggestion
    stop(
      "You have supplied both \"file\" and \"sink\" arguments. ",
      "Please supply only one of them.",
      call. = FALSE
    )
```

dragosmg replied:
Done

ianmcook commented on r/R/csv.R:
```diff
@@ -623,9 +638,58 @@ readr_to_csv_convert_options <- function(na,
+  }
+
+  if (!missing(col_names) && !missing(include_header)) {
+    stop(
+      paste("You have supplied both \"col_names\" and \"include_header\"",
+            "arguments. Please supply only one of them."),
+      call. = FALSE
+    )
```
```suggestion
    stop(
      "You have supplied both \"col_names\" and \"include_header\" ",
      "arguments. Please supply only one of them.",
      call. = FALSE
    )
```

dragosmg replied:
Done

ianmcook commented on r/R/csv.R:
```diff
@@ -623,9 +638,58 @@ readr_to_csv_convert_options <- function(na,
+
+  # default value are considered missing by base R
+  if (missing(include_header) && !missing(col_names)) {
+    message(
+      paste(
+        "You have supplied a value for \"col_names\". This will overwrite",
+        "the value for the \"include_headers\" argument."))
```
`message()` pastes together its arguments (with no separator)
```suggestion
    message(
      "You have supplied a value for \"col_names\". This will overwrite ",
      "the value for the \"include_headers\" argument."
    )
```

dragosmg replied:
Done

jonkeane reviewed (commented):
Oops, I meant to send these yesterday, but didn't click the submit button. Sorry about that!

dragosmg commented on r/R/csv.R:
```diff
@@ -623,9 +638,57 @@ readr_to_csv_convert_options <- function(na,
+      "You have supplied a value for \"col_names\". This will overwrite ",
+      "the value for the \"include_headers\" argument.")
+    include_header <- col_names
+  }
```
I'm trying to figure what would be the best way to deal with this. `include_header` is `TRUE` by default. The main use case for the message was to warn when the user tries to pass a different / contradictory value for `col_names` which overwrites the default value for `include_header`. I could just make the message appear then `col_names == FALSE`.

dragosmg replied:
In the end I got rid of the message. Done!

dragosmg commented on r/R/csv.R:
```diff
@@ -415,6 +415,16 @@ CsvReadOptions$create <- function(use_threads = option_use_threads(),
 
+readr_to_csv_write_options <- function(include_header,
+                                       batch_size = 1024L) {
+  assert_that(is_integerish(batch_size, n = 1, finite = TRUE), batch_size > 0)
```
It will not pass the assertion, more specifically the `finite = TRUE` part.
```
> is_integerish(Inf, n = 1)
[1] TRUE 
> is_integerish(Inf, n = 1, finite = TRUE)
[1] FALSE
```

dragosmg replied:
If we disabled that check / assertion in both `readr_to_csv_write_options` and `CsvWriteOptions$create()`, we would surface the following error message from C++:
```
Error: Invalid: WriteOptions: batch_size must be at least 1: -2147483648
/Users/dragos/Documents/arrow/cpp/src/arrow/csv/writer.cc:291  options.Validate()
/Users/dragos/Documents/arrow/cpp/src/arrow/csv/writer.cc:433  MakeCSVWriter(output, table.schema(), options)
In addition: Warning message:
In csv___WriteOptions__initialize(list(include_header = include_header,  :
  NAs introduced by coercion to integer range
 ```

dragosmg replied:
Should we aim for a more informative error message here? 
```
> write_csv_arrow(my_tibble, csv_file, batch_size = Inf)
Error: is_integerish(x = batch_size, n = 1, finite = TRUE) is not TRUE
```
might not be the easiest to digest by a user encountering `rlang::is_integerish()` for the first time. We also offer no direction for further action.

jonkeane reviewed (approved):
Thanks for this + being patient with the feedback!
